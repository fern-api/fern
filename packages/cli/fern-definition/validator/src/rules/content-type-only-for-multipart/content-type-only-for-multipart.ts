import { isInlineRequestBody, parseFileUploadRequest } from "@fern-api/fern-definition-schema";

import { Rule, RuleViolation } from "../../Rule";

export const ContentTypeOnlyForMultipartRule: Rule = {
    name: "content-type-only-for-multipart",
    DISABLE_RULE: false,
    create: () => {
        return {
            definitionFile: {
                httpEndpoint: ({ endpoint }) => {
                    if (endpoint.request == null) {
                        return [];
                    }

                    const parsedFileUploadRequest = parseFileUploadRequest(endpoint.request);
                    if (parsedFileUploadRequest != null) {
                        return [];
                    }

                    if (
                        typeof endpoint.request !== "string" &&
                        endpoint.request.body != null &&
                        isInlineRequestBody(endpoint.request.body)
                    ) {
                        const violations: RuleViolation[] = [];
                        for (const [propertyName, propertyDeclaration] of Object.entries(
                            endpoint.request.body.properties ?? {}
                        )) {
                            if (typeof propertyDeclaration === "string") {
                                continue;
                            }
                            if (propertyDeclaration["content-type"] != null) {
                                violations.push({
                                    severity: "fatal",
                                    message: `${propertyName} has content-type, but the request is not multipart`
                                });
                            }
                        }
                        return violations;
                    }

                    return [];
                }
            }
        };
    }
};
