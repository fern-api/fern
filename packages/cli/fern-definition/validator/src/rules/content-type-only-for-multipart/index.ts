export { ContentTypeOnlyForMultipartRule } from "./content-type-only-for-multipart";
