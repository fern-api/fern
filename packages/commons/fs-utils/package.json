{
  "name": "@fern-api/fs-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/fern-api/fern.git",
    "directory": "packages/commons/fs-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "vitest --passWithNoTests --run",
    "test:update": "vitest --passWithNoTests --run -u",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-pattern=../../../.eslintignore",
    "lint:eslint:fix": "yarn lint:eslint --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/core-utils": "workspace:*",
    "@fern-api/path-utils": "workspace:*",
    "glob": "^11.0.0",
    "json-stream-stringify": "^3.1.4",
    "stream-json": "^1.8.0",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@fern-api/core-utils": "workspace:*",
    "@fern-api/path-utils": "workspace:*",
    "@fern-api/configs": "workspace:*",
    "@types/node": "18.15.3",
    "depcheck": "^1.4.7",
    "eslint": "^8.56.0",
    "prettier": "^3.4.2",
    "@trivago/prettier-plugin-sort-imports": "^5.2.1",
    "typescript": "5.7.2",
    "@types/jest": "^29.5.14",
    "@types/stream-json": "^1.7.7",
    "vitest": "^2.1.8"
  }
}