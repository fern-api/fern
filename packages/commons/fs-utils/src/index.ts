export { AbsoluteFilePath } from "./AbsoluteFilePath";
export { cwd } from "./cwd";
export { dirname } from "./dirname";
export { doesPathExist, doesPathExistSync } from "./doesPathExist";
export {
    getDirectoryContents,
    type Directory,
    type File,
    type FileOrDirectory,
    getDirectoryContentsForSnapshot
} from "./getDirectoryContents";
export { join } from "./join";
export { listFiles } from "./listFiles";
export { moveFile } from "./moveFile";
export { moveFolder } from "./moveFolder";
export { relative } from "./relative";
export { RelativeFilePath } from "./RelativeFilePath";
export { relativize } from "./relativize";
export { resolve } from "./resolve";
export { streamObjectToFile } from "./streamObjectToFile";
export { stringifyLargeObject } from "./stringifyLargeObject";
export { waitUntilPathExists } from "./waitUntilPathExists";
export { streamObjectFromFile } from "./streamObjectFromFile";
export {
    convertToOsPath,
    convertToFernHostAbsoluteFilePath,
    convertToFernHostRelativeFilePath
} from "./osPathConverter";
export { getAllFilesInDirectory } from "./getAllFilesInDirectory";
export { getFilename } from "./getFilename";
export { isURL } from "./isUrl";
export { isCI } from "./isCI";
export { splitPath } from "./splitPath";
