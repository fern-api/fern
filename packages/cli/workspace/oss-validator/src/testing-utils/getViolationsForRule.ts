import stripAnsi from "strip-ansi";

import { AbsoluteFilePath } from "@fern-api/fs-utils";
import { OSSWorkspace } from "@fern-api/lazy-fern-workspace";
import { CONSOLE_LOGGER } from "@fern-api/logger";
import { createMockTaskContext } from "@fern-api/task-context";
import { loadAPIWorkspace } from "@fern-api/workspace-loader";

import { Rule } from "../Rule";
import { ValidationViolation } from "../ValidationViolation";
import { runRulesOnOSSWorkspace } from "../validateOSSWorkspace";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToWorkspace: AbsoluteFilePath;
        cliVersion?: string;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToWorkspace,
    cliVersion
}: getViolationsForRule.Args): Promise<ValidationViolation[]> {
    const context = createMockTaskContext();
    const result = await loadAPIWorkspace({
        absolutePathToWorkspace,
        context,
        cliVersion: "0.0.0",
        workspaceName: undefined
    });

    if (!result.didSucceed) {
        throw new Error("API workspace failed to load");
    }

    if (!(result.workspace instanceof OSSWorkspace)) {
        throw new Error("Expected an OSS workspace but got a different type");
    }

    const violations = await runRulesOnOSSWorkspace({
        workspace: result.workspace as OSSWorkspace,
        context,
        rules: [rule]
    });

    return violations.map((violation) => ({
        ...violation,
        message: stripAnsi(violation.message)
    }));
}
