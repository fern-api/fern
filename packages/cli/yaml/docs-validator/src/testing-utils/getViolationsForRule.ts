import stripAnsi from "strip-ansi";

import { filterOssWorkspaces } from "@fern-api/docs-resolver";
import { AbsoluteFilePath } from "@fern-api/fs-utils";
import { loadProjectFromDirectory } from "@fern-api/project-loader";
import { createMockTaskContext } from "@fern-api/task-context";

import { Rule } from "../Rule";
import { ValidationViolation } from "../ValidationViolation";
import { runRulesOnDocsWorkspace } from "../validateDocsWorkspace";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToFernDirectory: AbsoluteFilePath;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToFernDirectory
}: getViolationsForRule.Args): Promise<ValidationViolation[]> {
    const context = createMockTaskContext();
    const project = await loadProjectFromDirectory({
        absolutePathToFernDirectory,
        context,
        cliVersion: "0.0.0",
        defaultToAllApiWorkspaces: true,
        commandLineApiWorkspace: undefined,
        cliName: "fern"
    });

    if (project.docsWorkspaces == null) {
        throw new Error("Expected docs workspace to be present, but found none");
    }

    const violations = await runRulesOnDocsWorkspace({
        workspace: project.docsWorkspaces,
        context,
        rules: [rule],
        apiWorkspaces: project.apiWorkspaces,
        ossWorkspaces: await filterOssWorkspaces(project)
    });

    return violations.map((violation) => ({
        ...violation,
        message: stripAnsi(violation.message)
    }));
}
