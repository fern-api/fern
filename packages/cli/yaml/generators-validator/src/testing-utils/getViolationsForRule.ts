import stripAnsi from "strip-ansi";

import { loadGeneratorsConfiguration } from "@fern-api/configuration-loader";
import { AbsoluteFilePath } from "@fern-api/fs-utils";
import { LazyFernWorkspace } from "@fern-api/lazy-fern-workspace";
import { CONSOLE_LOGGER } from "@fern-api/logger";
import { createMockTaskContext } from "@fern-api/task-context";

import { Rule } from "../Rule";
import { ValidationViolation } from "../ValidationViolation";
import { runRulesOnWorkspace } from "../validateGeneratorsWorkspace";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToWorkspace: AbsoluteFilePath;
        cliVersion?: string;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToWorkspace,
    cliVersion
}: getViolationsForRule.Args): Promise<ValidationViolation[]> {
    const context = createMockTaskContext();

    const lazyWorkspace = new LazyFernWorkspace({
        absoluteFilePath: absolutePathToWorkspace,
        generatorsConfiguration: await loadGeneratorsConfiguration({
            absolutePathToWorkspace,
            context
        }),
        context,
        cliVersion: cliVersion ?? "0.0.0",
        workspaceName: undefined
    });
    const fernWorkspace = await lazyWorkspace.toFernWorkspace({ context });

    const violations = await runRulesOnWorkspace({
        workspace: fernWorkspace,
        logger: CONSOLE_LOGGER,
        rules: [rule]
    });

    return violations.map((violation) => ({
        ...violation,
        message: stripAnsi(violation.message)
    }));
}
