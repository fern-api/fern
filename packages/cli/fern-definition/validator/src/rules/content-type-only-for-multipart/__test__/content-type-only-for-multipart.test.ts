import { AbsoluteFilePath, RelativeFilePath, join } from "@fern-api/fs-utils";

import { getViolationsForRule } from "../../../testing-utils/getViolationsForRule";
import { ContentTypeOnlyForMultipartRule } from "../content-type-only-for-multipart";

describe("compatible-ir-versions", () => {
    it("simple failure", async () => {
        const violations = await getViolationsForRule({
            rule: ContentTypeOnlyForMultipartRule,
            absolutePathToWorkspace: join(
                AbsoluteFilePath.of(__dirname),
                RelativeFilePath.of("fixtures"),
                RelativeFilePath.of("simple")
            )
        });

        expect(violations).toMatchSnapshot();
    });
});
