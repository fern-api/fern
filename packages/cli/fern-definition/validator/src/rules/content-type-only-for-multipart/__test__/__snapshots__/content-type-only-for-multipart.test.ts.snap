// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compatible-ir-versions > simple failure 1`] = `
[
  {
    "message": "baz has content-type, but the request is not multipart",
    "nodePath": [
      "service",
      "endpoints",
      "json",
    ],
    "relativeFilepath": "simple.yml",
    "severity": "fatal",
  },
]
`;
